Ssa: Trim the use-use relation to skip irrelevant nodes

aschackmull opened the pull request:
This PR contains 3 tweaks to the shared SSA use-use step relation in the data flow integration module. Each of them trims the step relation in order to generate fewer nodes and fewer edges.
* `WriteDefinition`s are skipped for Java. There should be no need to include an extra node in the step from the RHS of an assignment to the first use of the variable, but some languages may depend on the flow out of definitions for now, so so far this is opt-in only.
* Synthetic reads on the input edges to phi nodes are necessary for proper BarrierGuards. But it's only a fraction of them that are actually potentially needed by any guard, so we can restrict their creation quite a bit. Furthermore, I've added a hook to allow certain SSA usages to skip these nodes entirely if BarrierGuards are not going to be used. I expect to use this option in the VariableCapture use-case.
* Finally, phi nodes only exist as intermediate nodes to prevent blow-ups in the number of edges, so if the successor is unique, we can safely skip the phi node.

geoffw0 commented:
Looking at the DCA runs:
- CPP has 4 new and 4 lost results for `cpp/invalid-pointer-deref`.  They're not just moved results.  I'm not sure what's going on with these.
- Rust shows a large increase in data flow inconsistencies, which should be understood before we merge this.  There's also a small (but surprising) improvement to taint reach and a 3.8% analysis slowdown (I'm not sure if either is significant).
- Swift hasn't been run (yet)

I'm a bit surprised because from the PR description I wasn't expecting to see changes in results.

aschackmull commented:
I've restarted dca after fixing a performance bug, the 3.8% slowdown on Rust is now 0.1% instead. As for the data flow inconsistencies (for e.g. Rust), these were fixed by the commit "SSA: Skip identity steps".

There should be no need to run dca for Swift, since Swift doesn't use the Data Flow Integration module.

As I mentioned on slack, result changes should not occur, but they do for C++ due to the somewhat ad-hoc `DataFlow::flowsToBackEdge` barrier, which breaks the SSA abstraction boundary somewhat.

aschackmull commented:
I've fixed the caching aspect for Java, and introduced a hook for C++, such that the `flowsToBackEdge` barrier ought to keep working as-is. For C++ this shaves a bit of the potential improvement, but we still get a somewhat decent node/edge reduction.

MathiasVP reviewed (commented):
C++ 👍 I'll leave the shared parts to someone else, as I don't really have time to look at this now, unfortunately 😭

hvitved commented on shared/ssa/codeql/ssa/Ssa.qll:
```diff
@@ -1533,6 +1539,29 @@ module Make<LocationSig Location, InputSig<Location> Input> {
+      (
+        exists(DfInput::Guard g | g.controlsBranchEdge(input, phi.getBasicBlock(), _))
+        or
+        exists(BasicBlock prev |
```
I think this could deserve a comment with an example

hvitved commented on shared/ssa/codeql/ssa/Ssa.qll:
```diff
@@ -1892,11 +1946,13 @@ module Make<LocationSig Location, InputSig<Location> Input> {
+        phi = nodeFrom.(SsaInputNodeImpl).getPhi() and
+        phi.definesAt(v, bbPhi, _, _) and
+        isUseStep = false and
+        if phiHasUniqNextNode(phi)
+        then flowFromRefToNode(v, bbPhi, -1, nodeTo)
+        else nodeTo.(SsaDefinitionExtNodeImpl).getDefExt() = phi
```
Pull out into separate predicate to avoid code duplication?

hvitved reviewed (commented):
LGTM, a few questions.

aschackmull commented on shared/ssa/codeql/ssa/Ssa.qll:
```diff
@@ -1495,6 +1495,13 @@ module Make<LocationSig Location, InputSig<Location> Input> {
     predicate guardControlsBlock(Guard guard, BasicBlock bb, boolean branch);
+
+    /**
+     * Holds if `WriteDefinition`s should be included as an intermediate node
```
Yes and no. Flipping this switch will remove all _certain_ writes from the use-use steps. It will also cause us to step over _uncertain_ write nodes for the RHS-to-first-use steps, but those nodes will remain in the graph as stepping stones from prior uses. It's potentially possible to skip over _uncertain_ writes when stepping from prior uses, but that requires some additional analysis to ensure that we only do so when it won't cause blowups, and the potential gain seemed so small that I didn't bother with that particular graph reduction.

aschackmull commented on shared/ssa/codeql/ssa/Ssa.qll:
```diff
@@ -1533,6 +1539,29 @@ module Make<LocationSig Location, InputSig<Location> Input> {
+    private predicate relevantPhiInputNode(SsaPhiExt phi, BasicBlock input) {
+      DfInput::supportBarrierGuardsOnPhiEdges() and
+      // If the input isn't explicitly read then a guard cannot check it.
+      exists(DfInput::getARead(getAPhiInputDef(phi, input))) and
```
Yes.
